features/users/users_controller.rb

```diff
@@ -0,0 +1,73 @@
+require "set"
+
+require_relative "../../shared/authenticated_controller_base"
+require_relative "../../services/services"
+
+module FastlaneCI
+  # A CRUD controller to manage users
+  class UsersController < AuthenticatedControllerBase
+    HOME = "/users"
+
+    get HOME do
+      locals = { title: "Users" }
+      erb(:users, locals: locals, layout: FastlaneCI.default_layout)
+    end
+
+    # When the `/users/create` form is submitted:
+    #
+    # - creates a user if the locals are valid
+    post "#{HOME}/create" do
+      if valid_params?(params, users_params)
+        Services.user_service.create_user!(
+          id: params[:id],
+          email: params[:email],
+          password: params[:password]
+        )
+      end
+
+      redirect(HOME)
+    end
+
+    # Updates a user existing in the configuration repository `users.json`
+    post "#{HOME}/update" do
+      if valid_params?(params, users_params)
+        new_user = User.new(
+          id: params[:id],
+          email: params[:email],
+          password: params[:password]
+        )
+
+        Services.user_service.update_user!(new_user)
+      end
+
+      redirect(HOME)
+    end
+
+    private
+
+    #####################################################
+    # @!group Data: View-specific data
+    #####################################################
+
+    # @return [Array[User]]
+    def users
+      Services.user_service.user_data_source.users
+    end
+
+    # Empty user object for new user form
+    #
+    # @return [User]
+    def new_user
+      @new_user ||= User.new(provider_credentials: [GitHubProviderCredential.new])
```

Comment:
Also, does this support adding multiple new users?

Reply:
Yep! So the way this works is in the view form, this `new_user` is [passed into the form partial](https://github.com/AndrewMcBurney/ci/blob/users_resource/features/users/views/users.erb#L54) to populate the [`create` form](https://github.com/AndrewMcBurney/ci/blob/users_resource/features/users/views/forms/_users.erb#L1).

It's not actually the same `User` that's created in the `post "#{HOME}/create"` action, which is why I memoized it (to avoid creating a new user in memory each time a `/users` view is rendered). I agree the name is kinda confusing, so maybe something like naming this method `blank_form_user` as well as a descriptive comment about where it's used would clear things up a bit better.

My logic for this was that I'd be able to use the user form for both the `create` and `update` actions.

Reply:
Oh, that's great. I like that name too 💯